File: Model/Source/Measurement.vb
Imports System.ComponentModel
Imports Wells.Model.ReflectionExtension

Public Class Measurement
    Inherits BusinessObject
    Implements IComparable(Of Measurement)

    <Browsable(False)>
    ReadOnly Property WellId As String
        Get
            Return Well?.Id
        End Get
    End Property

    Private _WellName As String
    <DisplayName("Pozo"), Browsable(True)>
    Property WellName As String
        Get
            If Well IsNot Nothing Then
                Return Well.Name
            Else
                Return _WellName
            End If
        End Get
        Set
            _WellName = Value
        End Set
    End Property

    <DisplayName("Fecha"), Browsable(True)>
    ReadOnly Property RealDate As Date
        Get
            Return Date.ParseExact(SampleDate, "dd/MM/yyyy", Nothing)
        End Get
    End Property

    <Browsable(False)>
    Property SampleDate As String

    <DisplayName("Profundidad FLNA"), Browsable(True)>
    Property FLNADepth As Double

    <DisplayName("Profundidad Agua"), Browsable(True)>
    Property WaterDepth As Double

    <DisplayName("Espesor FLNA"), Browsable(True)>
    ReadOnly Property FLNAThickness As Double
        Get
            If HasFLNA AndAlso HasWater Then
                Return WaterDepth - FLNADepth
            Else
                Return NullNumericValue
            End If
        End Get
    End Property

    <DisplayName("Cota FLNA"), Browsable(True)>
    ReadOnly Property FLNAElevation As Double
        Get
            If HasFLNA Then
                If Well?.HasHeight Then
                    Return Well?.Z + Well?.Height - FLNADepth
                Else
                    Return Well?.Z - FLNADepth
                End If
            End If
            Return NullNumericValue
        End Get
    End Property

    <DisplayName("Cota Agua"), Browsable(True)>
    ReadOnly Property WaterElevation As Double
        Get
            If HasWater Then
                If Well?.HasHeight Then
                    Return Well?.Z + Well?.Height - WaterDepth
                Else
                    Return Well?.Z - WaterDepth
                End If
            End If
            Return NullNumericValue
        End Get
    End Property

    <DisplayName("Caudal"), Browsable(True)>
    Property Caudal As Double

    <DisplayName("Observaciones"), Browsable(True)>
    Property Comment As String

    <Browsable(False)>
    Property Well As Well

    <Browsable(False)>
    ReadOnly Property HasFLNA As Boolean
        Get
            Return FLNADepth <> NullNumericValue
        End Get
    End Property

    <Browsable(False)>
    ReadOnly Property HasWater As Boolean
        Get
            Return WaterDepth <> NullNumericValue
        End Get
    End Property

    Public Overrides Function ToString() As String
        Return SampleDate
    End Function

    Public Function CompareTo(other As Measurement) As Integer Implements IComparable(Of Measurement).CompareTo
        If RealDate > other.RealDate Then Return -1
        If RealDate = other.RealDate Then Return 0
        Return 1
    End Function

    Shared ReadOnly Property Propeties As New Dictionary(Of String, String) From {
        {GetDisplayName(Of Measurement)(NameOf(Measurement.FLNADepth)), NameOf(Measurement.FLNADepth)},
        {GetDisplayName(Of Measurement)(NameOf(Measurement.WaterDepth)), NameOf(Measurement.WaterDepth)},
        {GetDisplayName(Of Measurement)(NameOf(Measurement.Caudal)), NameOf(Measurement.Caudal)},
        {GetDisplayName(Of Measurement)(NameOf(Measurement.FLNAThickness)), NameOf(Measurement.FLNAThickness)},
        {GetDisplayName(Of Measurement)(NameOf(Measurement.WaterElevation)), NameOf(Measurement.WaterElevation)},
        {GetDisplayName(Of Measurement)(NameOf(Measurement.FLNAElevation)), NameOf(Measurement.FLNAElevation)}}
End Class

File: Model/Model/src/Measurement.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Reflection;
using Wells.BaseModel.Models;

namespace Wells.Model
{
    public class Measurement : BusinessObject
    {
        public Measurement() { }

        #region Properties
        [Required]
        [Browsable(false)]
        public string WellId { get; set; }


        [Browsable(true), DisplayName("Pozo")]
        public string WellName => Well?.Name;


        [DisplayName("Fecha"), Browsable(true)]
        public DateTime Date { get; set; }


        [DisplayName("Profundidad FLNA (m)"), Browsable(true)]
        public double FlnaDepth { get; set; }


        [DisplayName("Profundidad Agua (m)"), Browsable(true)]
        public double WaterDepth { get; set; }


        [Browsable(false)]
        public bool HasWater => WaterDepth != NumericNullValue;


        [Browsable(false)]
        public bool HasFlna => FlnaDepth != NumericNullValue;

        [DisplayName("Espesor FLNA (m)"), Browsable(true)]
        public double FlnaThickness
        {
            get
            {
                if (HasFlna && HasWater) { return WaterDepth - FlnaDepth; }
                return NumericNullValue;
            }
        }


        [DisplayName("Cota FLNA (m)"), Browsable(true)]
        public double FlnaElevation
        {
            get
            {
                if (HasFlna && Well != null)
                {
                    if (Well.HasHeight && Well.HasZ) { return Well.Z + Well.Height - FlnaDepth; }
                    else if (Well.HasZ) { return Well.Z - FlnaDepth; }
                    return -FlnaDepth;
                }
                return NumericNullValue;
            }
        }


        [DisplayName("Cota Agua (m)"), Browsable(true)]
        public double WaterElevation
        {
            get
            {
                if (HasWater && Well != null)
                {
                    if (Well.HasHeight && Well.HasZ) { return Well.Z + Well.Height - WaterDepth; }
                    else if (Well.HasZ) { return Well.Z - WaterDepth; }
                    return -WaterDepth;
                }
                return NumericNullValue;
            }
        }

        [DisplayName("Observaciones"), Browsable(true)]
        public string Comment { get; set; }
        #endregion


        #region Lazy-Load Properties
        /// <summary>
        /// The parent well.
        /// </summary>
        [ForeignKey("WellId")]
        [Browsable(false)]
        public virtual Well Well { get; set; }

        #endregion


        public override string ToString()
        {
            return Date.ToString(DateFormat);
        }

        public override int CompareTo(IBusinessObject other)
        {
            return Date.CompareTo((other as Measurement).Date);
        }


        [Browsable(false)]
        public static Dictionary<string, PropertyInfo> Properties
        {
            get
            {
                return GetBrowsableProperties(typeof(Measurement));
            }
        }

        [Browsable(false)]
        public static Dictionary<string, PropertyInfo> DoubleProperties
        {
            get
            {
                return GetDoubleBrowsableProperties(typeof(Measurement));
            }
        }

        public static string GetDisplayName(string propertyName)
        {
            return GetDisplayName(typeof(Measurement), propertyName);
        }
    }
}
